File: src/api/MixServer.Domain/FileExplorer/Models/NodePath.cs
namespace MixServer.Domain.FileExplorer.Models;

public record NodePath(string RootPath, string RelativePath)
{
    public string AbsolutePath => Path.Join(RootPath, RelativePath);
    public string FileName => string.IsNullOrWhiteSpace(InternalFileName) ? AbsolutePath : InternalFileName;
    public NodePath Parent => this with { RelativePath = Path.GetDirectoryName(RelativePath) ?? string.Empty };
    
    public string Extension => Path.GetExtension(RelativePath);
    
    public bool IsRoot => string.IsNullOrWhiteSpace(RootPath) && string.IsNullOrWhiteSpace(RelativePath);
    
    public bool IsRootChild => !string.IsNullOrWhiteSpace(RootPath) && string.IsNullOrWhiteSpace(RelativePath);

    public override string ToString()
    {
        return AbsolutePath;
    }

    public bool IsEqualTo(NodePath? other)
    {
        if (other is null)
        {
            return false;
        }
        
        return RootPath == other.RootPath && RelativePath == other.RelativePath;
    }
    
    private string InternalFileName => Path.GetFileName(AbsolutePath);
}

File: src/clients/mix-server-client/src/app/main-content/file-explorer/models/node-path.ts
export class NodePathHeader {
  constructor(public rootPath: string,
              public relativePath: string)
  {
  }

  // this is not the absolute path, but the key to identify the node
  // As the client should not need the logic to know different platforms path separators
  public get key(): string {
    return this.rootPath + ";" + this.relativePath;
  }

  public get empty(): boolean {
    return this.rootPath === '' && this.relativePath === '';
  }

  public static get Default(): NodePathHeader {
    return new NodePathHeader('', '');
  }

  public copy(): NodePathHeader {
    return new NodePathHeader(this.rootPath, this.relativePath);
  }

  public isEqual(nodePathParent: NodePathHeader | null | undefined): boolean {
    if (!nodePathParent) {
      return false;
    }

    return this.rootPath === nodePathParent.rootPath && this.relativePath === nodePathParent.relativePath;
  }
}

export class NodePath extends NodePathHeader {
  constructor(rootPath: string,
              relativePath: string,
              public fileName: string,
              public absolutePath: string,
              public extension: string,
              public parent: NodePathHeader,
              public isRoot: boolean,
              public isRootChild: boolean) {
    super(rootPath, relativePath);
  }

  public static override get Default(): NodePath {
    return new NodePath('', '', '', '', '', NodePathHeader.Default, false, false);
  }

  public override copy(): NodePath {
    return new NodePath(this.rootPath, this.relativePath, this.fileName, this.absolutePath, this.extension, this.parent.copy(), this.isRoot, this.isRootChild);
  }
}
